frontend/src/metabase/visualizations/lib/LineAreaBarRenderer.js

```diff
@@ -224,6 +224,70 @@ function getDimensionsAndGroupsAndUpdateSeriesDisplayNamesForStackedChart(
   return { dimension, groups };
 }
 
+// ASSERT(datas.length === 1)
+// ASSERT(datas[0].length > 0)
+function getDimensionsAndGroupsForWaterfallChart(props, originalDatas, warn) {
+  const datas = originalDatas.slice();
+  datas.push(datas[0].map(k => k.slice())); // negatives
+  datas.push(datas[0].map(k => k.slice())); // positives
+
+  /*
+  A waterfall chart is essentially a stacked bar chart.
+  It consists of the following (from the topmost):
+  - the "green bar", when the current value is greater than the previous one
+  - the "red bar", when the current value is smaller than the previous one
+  - the "invisible beam" supporting either the positive or negative bar
+
+  Note the green and red bars are mutually exclusive (i.e. if one has
+  a positive value, the other is zero). This is done so we need not have
+  conditional fill color.
+  */
+
+  const values = datas[0].map(d => d[1]);
+  const deltas = values.reduce(
+    (result, v) => {
+      result.deltas.push(v - result.last);
+      result.last = v;
+      return result;
+    },
+    { last: values[0], deltas: [] },
+  ).deltas;
+  const positives = deltas.map(v => (v > 0 ? v : 0));
+  positives[0] = values[0];
+  const negatives = deltas.map(v => (v < 0 ? -v : 0));
+  const beams = [0];
+  for (let i = 1; i < deltas.length; ++i) {
+    beams[i] = positives[i] > 0 ? values[i - 1] : values[i - 1] - negatives[i];
+  }
+  for (let k = 0; k < values.length; ++k) {
+    datas[0][k][1] = beams[k];
+    datas[1][k][1] = negatives[k];
+    datas[2][k][1] = positives[k];
+  }
```

Comment:
This dataset switcheroo is rather messy, probably needs more samples and clarification.
OTOH, an alternative, likely cleaner approach is to do it on the crossfilter itself. Thoughts?

Reply:
As a more concrete example, here's a screenshot from the product doc:

![image](https://user-images.githubusercontent.com/691495/99099484-7efa5300-25a8-11eb-88c6-c23e7d2ac7e3.png)

I'm imagining the first three rows of data for this chart will be:

```
Armoires,222
Bath,60
Beds,-100
```
rather than
```
Armoires,222
Bath,282
Beds,182
```